<a href="https://colab.research.google.com/github/Ahmamaher/Accident-Data-Analysis/blob/main/Copy_of_Task.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Python Project Data - accident data**

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

# **Data Gathering**

In [ ]:
df = pd.read_csv("/content/accident data (1).csv")
df

Index Accident_Severity Accident Date   Latitude  \
0       200701BS64157           Serious    05-06-2019  51.506187   
1       200701BS65737           Serious    02-07-2019  51.495029   
2       200701BS66127           Serious    26-08-2019  51.517715   
3       200701BS66128           Serious    16-08-2019  51.495478   
4       200701BS66837            Slight    03-09-2019  51.488576   
...               ...               ...           ...        ...   
660674  201091NM01760            Slight    18-02-2022  57.374005   
660675  201091NM01881            Slight    21-02-2022  57.232273   
660676  201091NM01935            Slight    23-02-2022  57.585044   
660677  201091NM01964           Serious    23-02-2022  57.214898   
660678  201091NM02142           Serious    28-02-2022  57.575210   

                   Light_Conditions           District Area  Longitude  \
0             Darkness - lights lit  Kensington and Chelsea  -0.209082   
1                          Daylight  Kensington and Chelsea  -0.173647   
2       Darkness - lighting unknown  Kensington and Chelsea  -0.210215   
3                          Daylight  Kensington and Chelsea  -0.202731   
4             Darkness - lights lit  Kensington and Chelsea  -0.192487   
...                             ...                     ...        ...   
660674                     Daylight                Highland  -3.467828   
660675       Darkness - no lighting                Highland  -3.809281   
660676                     Daylight                Highland  -3.862727   
660677       Darkness - no lighting                Highland  -3.823997   
660678                     Daylight                Highland  -3.895673   

        Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
0                          1                   2                     Dry   
1                          1                   2             Wet or damp   
2                          1                   3                     Dry   
3                          1                   4                     Dry   
4                          1                   2                     Dry   
...                      ...                 ...                     ...   
660674                     2                   1                     Dry   
660675                     1                   1            Frost or ice   
660676                     1                   3            Frost or ice   
660677                     1                   2             Wet or damp   
660678                     1                   1             Wet or damp   

                 Road_Type Urban_or_Rural_Area     Weather_Conditions  \
0       Single carriageway               Urban     Fine no high winds   
1       Single carriageway               Urban  Raining no high winds   
2                      NaN               Urban                    NaN   
3       Single carriageway               Urban     Fine no high winds   
4                      NaN               Urban                    NaN   
...                    ...                 ...                    ...   
660674  Single carriageway               Rural     Fine no high winds   
660675  Single carriageway               Rural     Fine no high winds   
660676  Single carriageway               Rural     Fine no high winds   
660677  Single carriageway               Rural     Fine no high winds   
660678    Dual carriageway               Rural  Snowing no high winds   

                                Vehicle_Type  
0                                        Car  
1                                        Car  
2                      Taxi/Private hire car  
3       Bus or coach (17 or more pass seats)  
4                              Other vehicle  
...                                      ...  
660674                                   Car  
660675                                   Car  
660676                                   Car  
660677                 Motorcycle over 500cc  
660

# Data **Assessing**

In [ ]:
df.head()

Index Accident_Severity Accident Date   Latitude  \
0  200701BS64157           Serious    05-06-2019  51.506187   
1  200701BS65737           Serious    02-07-2019  51.495029   
2  200701BS66127           Serious    26-08-2019  51.517715   
3  200701BS66128           Serious    16-08-2019  51.495478   
4  200701BS66837            Slight    03-09-2019  51.488576   

              Light_Conditions           District Area  Longitude  \
0        Darkness - lights lit  Kensington and Chelsea  -0.209082   
1                     Daylight  Kensington and Chelsea  -0.173647   
2  Darkness - lighting unknown  Kensington and Chelsea  -0.210215   
3                     Daylight  Kensington and Chelsea  -0.202731   
4        Darkness - lights lit  Kensington and Chelsea  -0.192487   

   Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
0                     1                   2                     Dry   
1                     1                   2             Wet or damp   
2                     1                   3                     Dry   
3                     1                   4                     Dry   
4                     1                   2                     Dry   

            Road_Type Urban_or_Rural_Area     Weather_Conditions  \
0  Single carriageway               Urban     Fine no high winds   
1  Single carriageway               Urban  Raining no high winds   
2                 NaN               Urban                    NaN   
3  Single carriageway               Urban     Fine no high winds   
4                 NaN               Urban                    NaN   

                           Vehicle_Type  
0                                   Car  
1                                   Car  
2                 Taxi/Private hire car  
3  Bus or coach (17 or more pass seats)  
4                         Other vehicle

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 660679 entries, 0 to 660678
Data columns (total 14 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   Index                    660679 non-null  object 
 1   Accident_Severity        660679 non-null  object 
 2   Accident Date            660679 non-null  object 
 3   Latitude                 660654 non-null  float64
 4   Light_Conditions         660679 non-null  object 
 5   District Area            660679 non-null  object 
 6   Longitude                660653 non-null  float64
 7   Number_of_Casualties     660679 non-null  int64  
 8   Number_of_Vehicles       660679 non-null  int64  
 9   Road_Surface_Conditions  659953 non-null  object 
 10  Road_Type                656159 non-null  object 
 11  Urban_or_Rural_Area      660664 non-null  object 
 12  Weather_Conditions       646551 non-null  object 
 13  Vehicle_Type             660679 non-null  object 
dtypes: f

In [ ]:
df.isnull().sum()


Index                          0
Accident_Severity              0
Accident Date                  0
Latitude                      25
Light_Conditions               0
District Area                  0
Longitude                     26
Number_of_Casualties           0
Number_of_Vehicles             0
Road_Surface_Conditions      726
Road_Type                   4520
Urban_or_Rural_Area           15
Weather_Conditions         14128
Vehicle_Type                   0
dtype: int64

In [ ]:
df.describe()


Latitude      Longitude  Number_of_Casualties  Number_of_Vehicles
count  660654.000000  660653.000000         660679.000000       660679.000000
mean       52.553866      -1.431210              1.357040            1.831255
std         1.406922       1.383330              0.824847            0.715269
min        49.914430      -7.516225              1.000000            1.000000
25%        51.490690      -2.332291              1.000000            1.000000
50%        52.315641      -1.411667              1.000000            2.000000
75%        53.453452      -0.232869              1.000000            2.000000
max        60.757544       1.762010             68.000000           32.000000

In [ ]:
df.duplicated().sum()

19

## Quality **Issues**

**1-Error in the data type**

    1-The Index column must be an integer (int) but is an object
    2-The Accident Date column is an object instead of a datetime
2-Missing values


    1-Columns with missing values:
    2-Latitude (660654 out of 660679)
    3-Longitude (660653 out of 660679)
    4-Road_Surface_Conditions (659953 out of 660679)
    5-Road_Type (656159 out of 660679)
    6-Urban_or_Rural_Area (660664 out of 660679)
    7-Weather_Conditions (646551 out of 660679)
    
3-Duplicate Data:



# **Data Cleaning**



 ### **Delete** **duplicate**




In [ ]:
df.drop_duplicates(inplace=True)


### Dealing with incorrect data types


**Define:**
The Index column must be an integer (int) but is an object

**Code:**

In [93]:
df['Index'] = df['Index'].str.extract('(\d+)').astype(int)


# Test:

In [92]:
df.dtypes

Index                       object
Accident_Severity           object
Accident Date               object
Latitude                   float64
Light_Conditions            object
District Area               object
Longitude                  float64
Number_of_Casualties         int64
Number_of_Vehicles           int64
Road_Surface_Conditions     object
Road_Type                   object
Urban_or_Rural_Area         object
Weather_Conditions          object
Vehicle_Type                object
dtype: object

**Define:**
2-The Accident Date column is an object instead of a datetime


**Code**:

In [94]:
df['Accident Date'] = pd.to_datetime(df['Accident Date'], format='%d-%m-%Y')


**Test:**

In [95]:
df.dtypes

Index                               int64
Accident_Severity                  object
Accident Date              datetime64[ns]
Latitude                          float64
Light_Conditions                   object
District Area                      object
Longitude                         float64
Number_of_Casualties                int64
Number_of_Vehicles                  int64
Road_Surface_Conditions            object
Road_Type                          object
Urban_or_Rural_Area                object
Weather_Conditions                 object
Vehicle_Type                       object
dtype: object

**define:** Number_of_Casualties and number_of_vehicles their type Float64 but they are correct numbers

In [96]:
df['Number_of_Casualties'] = df['Number_of_Casualties'].astype(int)
df['Number_of_Vehicles'] = df['Number_of_Vehicles'].astype(int)


### **Missing values**

**Define:**
1-Fill out the lost values ​​with the most common values ​​(Mode) of the classified columns

2-Fill in the lost values ​​in the arithmetic middle of the digital columns


**Code:**

In [97]:
df['Road_Surface_Conditions'].fillna(df['Road_Surface_Conditions'].mode()[0], inplace=True)
df['Road_Type'].fillna(df['Road_Type'].mode()[0], inplace=True)
df['Weather_Conditions'].fillna(df['Weather_Conditions'].mode()[0], inplace=True)


<ipython-input-97-53a52566a8e3>:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['Road_Surface_Conditions'].fillna(df['Road_Surface_Conditions'].mode()[0], inplace=True)
<ipython-input-97-53a52566a8e3>:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[

**Code**:

In [98]:
df['Number_of_Casualties'].fillna(df['Number_of_Casualties'].mean(), inplace=True)
df['Number_of_Vehicles'].fillna(df['Number_of_Vehicles'].mean(), inplace=True)


<ipython-input-98-09b98810cc44>:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['Number_of_Casualties'].fillna(df['Number_of_Casualties'].mean(), inplace=True)
<ipython-input-98-09b98810cc44>:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].meth

**Test:**

In [99]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 660660 entries, 0 to 660678
Data columns (total 14 columns):
 #   Column                   Non-Null Count   Dtype         
---  ------                   --------------   -----         
 0   Index                    660660 non-null  int64         
 1   Accident_Severity        660660 non-null  object        
 2   Accident Date            660660 non-null  datetime64[ns]
 3   Latitude                 660635 non-null  float64       
 4   Light_Conditions         660660 non-null  object        
 5   District Area            660660 non-null  object        
 6   Longitude                660634 non-null  float64       
 7   Number_of_Casualties     660660 non-null  int64         
 8   Number_of_Vehicles       660660 non-null  int64         
 9   Road_Surface_Conditions  660660 non-null  object        
 10  Road_Type                660660 non-null  object        
 11  Urban_or_Rural_Area      660645 non-null  object        
 12  Weather_Conditions   

Define:Latitude and Longitude (geographical location):
Ways to compensate for lost values:

If accidents are within specific areas, they can be compensated by the average coordinates of the same area (District Area).

code

In [100]:
df['Latitude'].fillna(df.groupby('District Area')['Latitude'].transform('mean'), inplace=True)
df['Longitude'].fillna(df.groupby('District Area')['Longitude'].transform('mean'), inplace=True)


<ipython-input-100-615b41a7446b>:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['Latitude'].fillna(df.groupby('District Area')['Latitude'].transform('mean'), inplace=True)
<ipython-input-100-615b41a7446b>:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing

Test

In [101]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 660660 entries, 0 to 660678
Data columns (total 14 columns):
 #   Column                   Non-Null Count   Dtype         
---  ------                   --------------   -----         
 0   Index                    660660 non-null  int64         
 1   Accident_Severity        660660 non-null  object        
 2   Accident Date            660660 non-null  datetime64[ns]
 3   Latitude                 660660 non-null  float64       
 4   Light_Conditions         660660 non-null  object        
 5   District Area            660660 non-null  object        
 6   Longitude                660660 non-null  float64       
 7   Number_of_Casualties     660660 non-null  int64         
 8   Number_of_Vehicles       660660 non-null  int64         
 9   Road_Surface_Conditions  660660 non-null  object        
 10  Road_Type                660660 non-null  object        
 11  Urban_or_Rural_Area      660645 non-null  object        
 12  Weather_Conditions   

Define:Compensating the lost values ​​with the most frequent value (Mode).

Code

In [102]:
df['Urban_or_Rural_Area'].fillna(df['Urban_or_Rural_Area'].mode()[0], inplace=True)


<ipython-input-102-bcea9a9fe2fd>:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['Urban_or_Rural_Area'].fillna(df['Urban_or_Rural_Area'].mode()[0], inplace=True)


Test:

In [103]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 660660 entries, 0 to 660678
Data columns (total 14 columns):
 #   Column                   Non-Null Count   Dtype         
---  ------                   --------------   -----         
 0   Index                    660660 non-null  int64         
 1   Accident_Severity        660660 non-null  object        
 2   Accident Date            660660 non-null  datetime64[ns]
 3   Latitude                 660660 non-null  float64       
 4   Light_Conditions         660660 non-null  object        
 5   District Area            660660 non-null  object        
 6   Longitude                660660 non-null  float64       
 7   Number_of_Casualties     660660 non-null  int64         
 8   Number_of_Vehicles       660660 non-null  int64         
 9   Road_Surface_Conditions  660660 non-null  object        
 10  Road_Type                660660 non-null  object        
 11  Urban_or_Rural_Area      660660 non-null  object        
 12  Weather_Conditions   

# **Data Storing**

In [ ]:
df.to_csv('cleaned_data.csv', index=False)


In [ ]:
from google.colab import files
files.download('cleaned_data.csv')
